### PR TITLE
GEN-1261: add new CFactory

### PR DESCRIPTION
**Pull request by fedealconada**
## Motivation

Support cToken adapters via factory

## Solution

Add new hardhat task that deploys the new CFactory

## Implementer Checklist

- [ ]  [FIRST TIME ONLY] Review the [Solcurity](https://github.com/Rari-Capital/solcurity) standard
- [ ]  [OPTIONAL] Create a reference implementation (in python or JS)
- [x]  Create a short "Motivation" section for the PR
- [x]  Write a spec for the feature and put it in the PR description – basic function names and expected state transitions is OK
- [ ]  Check all of the new revert paths with concrete tests
- [ ]  Check all of the new storage slot writes with concrete tests – do not make the passing test case the trivial case!
- [ ]  Go line-by-line through the new code and ensure it has all been covered by concrete tests – since we don’t have a coverage engine for foundry yet, you can imagine how one might check that each different code path is covered
- [ ]  Simplify the implementation and spend some time trying to minimize gas costs
- [ ]  Write fuzz tests for the feature – try everything to break the implementation (is this function monotonically in/decreasing, should it always be less than something else, etc)
- [ ]  Capture bugs discovered during the above steps in concrete tests (regression tests)
- [ ]  Add integration tests – fuzz or concrete tests that test how the new code affects the entire system, either locally or with a mainnet fork
- [ ]  [OPTIONAL] Integrate the feature into the deployment scripts
- [ ]  [OPTIONAL] Add new sanity checks to the deployment scripts
- [ ]  Get the PR reviewed by at least two people

### CODE REVIEW

File: pkg/deployments/hardhat.addresses.js [hunk 1]
```diff
@@ -47,6 +47,11 @@ CUSDC_TOKEN.set("1", "0x39AA39c021dfbaE8faC545936693aC917d5E7563");
 CUSDC_TOKEN.set("111", "0x39AA39c021dfbaE8faC545936693aC917d5E7563");
 // TODO: Arbitrum
 
+const CUSDT_TOKEN = new Map();
+CUSDT_TOKEN.set("1", "0xf650C3d88D12dB855b8bf7D11Be6C55A4e07dCC9");
+CUSDT_TOKEN.set("111", "0xf650C3d88D12dB855b8bf7D11Be6C55A4e07dCC9");
+// TODO: Arbitrum
```
Comment (jparklev) [inline pkg/deployments/hardhat.addresses.js:53]: lol we can probably remove these comments for now – no upcoming plans for arbitrum atm

File: pkg/deployments/tasks/20220727-cfactory/input.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+const C_FACTORY_TARGETS = [
+  // { name: "cUSDT", series: [], address: CUSDT_TOKEN.get("1") },
+  { name: "cDAI", series: [], address: CDAI_TOKEN.get("1") },
+];
+
+const MAINNET_FACTORIES = [
+  {
+    contractName: "CFactory",
+    adapterContract: "CAdapter",
+    ifee: ethers.utils.parseEther("0.0010"),
```
Comment (jparklev) [inline pkg/deployments/tasks/20220727-cfactory/input.js:14]: Where did this issuance fee come from?

File: pkg/deployments/tasks/20220727-cfactory/index.js [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+      await (await divider.setIsTrusted(factoryAddress, true)).wait();
+
+      await hre.network.provider.request({
+        method: "hardhat_stopImpersonatingAccount",
+        params: [senseAdminMultisigAddress],
+      });
+
+      console.log("\n-------------------------------------------------------");
+      console.log(`Deploy adapters for: ${factoryContractName}`);
+      for (let t of targets) {
```
Comment (jparklev) [inline pkg/deployments/tasks/20220727-cfactory/index.js:93]: This is like a sanity check for local forks?

**Review by jparklev — APPROVED** [review]
👍 💻

File: pkg/deployments/tasks/20220727-cfactory/input.js (hunk 2 above)
Comment (jparklev) [reply pkg/deployments/tasks/20220727-cfactory/input.js:14]: Nvm – see it in linear now!

File: pkg/deployments/hardhat.addresses.js (hunk 1 above)
Comment (fedealconada) [reply pkg/deployments/hardhat.addresses.js:53]: hahah, true!

File: pkg/deployments/tasks/20220727-cfactory/index.js (hunk 3 above)
Comment (fedealconada) [reply pkg/deployments/tasks/20220727-cfactory/index.js:93]: yep, only for that

File: pkg/core/src/adapters/implementations/compound/CAdapter.sol [hunk 4]
```diff
@@ -101,7 +101,7 @@ contract CAdapter is BaseAdapter, Crop {
         address _reward
     ) Crop(_divider, _reward) BaseAdapter(_divider, _target, _underlying, _ifee, _adapterParams) {
         isCETH = _target == CETH;
-        ERC20(_underlying).approve(_target, type(uint256).max);
+        ERC20(_underlying).safeApprove(_target, type(uint256).max);
```
Comment (abylaw) [inline pkg/core/src/adapters/implementations/compound/CAdapter.sol:104]: for my own education, why use `safeApprove` here vs regular `approve`?

File: pkg/deployments/tasks/20220727-cfactory/index.js [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+      log: true,
+    });
+
+    console.log(`${factoryContractName} deployed to ${factoryAddress}`);
+
+    // verify on etherscan
+    if (chainId !== "111") {
+      console.log("\n-------------------------------------------------------");
+      await verifyOnEtherscan(factoryAddress, [divider.address, factoryParams]);
+    }
```
Comment (abylaw) [inline pkg/deployments/tasks/20220727-cfactory/index.js]: nit: connect this w/ if below into one `if..else`?

File: pkg/deployments/tasks/20220727-cfactory/index.js [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+      args: [divider.address, factoryParams, reward],
+      log: true,
+    });
+
+    console.log(`${factoryContractName} deployed to ${factoryAddress}`);
+
+    // verify on etherscan
+    if (chainId !== "111") {
+      console.log("\n-------------------------------------------------------");
+      await verifyOnEtherscan(factoryAddress, [divider.address, factoryParams]);
```
Comment (abylaw) [inline pkg/deployments/tasks/20220727-cfactory/index.js:64]: just to double-check/bring it up, we're now verifying on goerli as well? that'd be super useful for dev.

**Review by abylaw — APPROVED** [review]
LGTM~

File: pkg/deployments/tasks/20220727-cfactory/index.js (hunk 6 above)
Comment (fedealconada) [reply pkg/deployments/tasks/20220727-cfactory/index.js:64]: yes!! not sure why we were not doing it lol

Comment (abylaw) [reply pkg/deployments/tasks/20220727-cfactory/index.js:64]: hehe yay

**Pull request merged**